lib/membrane_mp4/muxer/isom.ex

```diff
@@ -0,0 +1,196 @@
+    state = Map.update!(state, :pad_order, &[pad_ref | &1])
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_caps(Pad.ref(:input, pad_ref), %Membrane.MP4.Payload{} = caps, _ctx, state) do
+    track =
+      caps
+      |> Map.take([:width, :height, :content, :timescale])
+      |> Track.new()
+
+    state = put_in(state, [:pad_to_track, pad_ref], track)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_demand(:output, _size, :buffers, _ctx, state) do
+    next_ref = hd(state.pad_order)
+
+    {{:ok, demand: {Pad.ref(:input, next_ref), 1}}, state}
+  end
+
+  @impl true
+  def handle_process(Pad.ref(:input, pad_ref), buffer, _ctx, state) do
+    state =
+      state
+      |> update_in([:pad_to_track, pad_ref], &Track.store_sample(&1, buffer))
+      |> maybe_flush_chunk(pad_ref)
+
+    {{:ok, redemand: :output}, state}
+  end
+
+  @impl true
+  def handle_end_of_stream(Pad.ref(:input, pad_ref), _ctx, state) do
+    state =
+      state
+      |> do_flush_chunk(pad_ref)
+      |> Map.update!(:pad_order, &List.delete(&1, pad_ref))
+
+    if length(state.pad_order) > 0 do
+      {{:ok, redemand: :output}, state}
+    else
+      mp4 = serialize(state)
+
+      {{:ok, buffer: {:output, %Buffer{payload: mp4}}, end_of_stream: :output}, state}
+    end
+  end
+
+  defp maybe_flush_chunk(state, pad_ref) do
+    use Ratio, comparison: true
+    track = get_in(state, [:pad_to_track, pad_ref])
+
+    if Track.current_chunk_duration(track) >= state.chunk_duration do
+      do_flush_chunk(state, pad_ref)
+    else
+      state
+    end
+  end
+
+  defp do_flush_chunk(state, pad_ref) do
+    track = get_in(state, [:pad_to_track, pad_ref])
+
+    {chunk, track} = Track.flush_chunk(track, chunk_offset(state))
+
+    state
+    |> Map.update!(:media_data, &(&1 <> chunk))
+    |> put_in([:pad_to_track, pad_ref], track)
+    |> Map.update!(:pad_order, &shift_left/1)
+  end
+
+  defp serialize(state) do
+    mdat = [mdat: %{content: state.media_data}]
+    moov = state.pad_to_track |> Map.values() |> MovieBox.assemble()
+
+    if state.fast_start do
+      [@ftyp, fast_start_moov(moov), mdat]
+    else
+      [@ftyp, mdat, moov]
+    end
+    |> Enum.map(&Container.serialize!/1)
+    |> Enum.reduce(&(&2 <> &1))
+  end
+
+  defp fast_start_moov(moov) do
+    moov_size = moov |> Container.serialize!() |> byte_size()
+    moov_children = get_in(moov, [:moov, :children])
+
+    # updates all `trak` boxes by adding `moov_size` to the offset of each chunk in its sample table
+    traks_with_offset =
+      moov_children
+      |> Keyword.get_values(:trak)
+      |> Enum.map(fn trak ->
+        Container.update_box(
+          trak.children,
+          [:mdia, :minf, :stbl, :stco],
+          [:fields, :entry_list],
+          &Enum.map(&1, fn %{chunk_offset: offset} -> %{chunk_offset: offset + moov_size} end)
+        )
+      end)
+      |> Enum.reduce([], &([trak: %{children: &1, fields: %{}}] ++ &2))
```

Comment:
I'm not a big fan of `reduce` as it's the most general of all Enum functions. Here I'd simply suggest using `Enum.map`
```suggestion
      |> Enum.map(&{:trak, %{children: &1, fields: %{}}})
```